amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+use self::internal::{Backtrace, HAS_BACKTRACE};
+use std::borrow::Cow;
+use std::env;
+use std::error;
+use std::ffi;
+use std::fmt;
+use std::result;
+use std::sync::atomic;
+
+/// Private
+enum ErrorImpl {
+    Message(Cow<'static, str>),
+    Error(Box<dyn error::Error + Send + Sync>),
+}
+
+impl fmt::Debug for ErrorImpl {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorImpl::Message(ref message) => fmt::Debug::fmt(message, fmt),
+            ErrorImpl::Error(ref e) => fmt::Debug::fmt(e, fmt),
+        }
+    }
+}
+
+impl From<&'static str> for ErrorImpl {
+    fn from(message: &'static str) -> Self {
+        ErrorImpl::Message(Cow::from(message))
+    }
+}
+
+impl From<String> for ErrorImpl {
+    fn from(message: String) -> Self {
+        ErrorImpl::Message(Cow::from(message))
+    }
+}
+
+impl From<Box<dyn error::Error + Send + Sync>> for ErrorImpl {
+    fn from(error: Box<dyn error::Error + Send + Sync>) -> Self {
+        ErrorImpl::Error(error)
+    }
+}
+
+/// The error type used by Amethyst.
+///
+/// Wraps error diagnostics like messages and other errors, and keeps track of causal chains and
+/// backtraces.
+pub struct Error {
+    error_impl: ErrorImpl,
+    source: Option<Box<Error>>,
+    backtrace: Option<Backtrace>,
```

Comment:
Do not include `Backtrace` at all if it is not enabled by compile feature (extra size for no needs)

Reply:
It's an important diagnostic tool. It will probably be enabled. See the linked issue at the top.

Reply:
I think what @DoumanAsh meant is to feature flag this field as to not make the struct any bigger than necessary.

Reply:
Yes, since we control it by feature, you should also exclude it from `Error` as it is overhead for users like me who wouldn't use backtrace

Reply:
Ah, sure!